[Diagnostics] Improve diagnostic when using == instead of = for defau…

vguerra opened the pull request:
…lt function argument.

Resolves [SR-11006](https://bugs.swift.org/browse/SR-11006).

vguerra committed:
[Diagnostics] Improve diagnostic when using == instead of = for default function argument.

theblixguy commented:
cc @rintaro

jrose-apple commented on lib/Parse/ParsePattern.cpp:
```diff
@@ -366,7 +366,13 @@ Parser::parseParameterClause(SourceLoc &leftParenLoc,
+        if (Tok.isBinaryOperator() && Tok.getText() == "==") {
+          diagnose(Tok,
+                   diag::expected_assignment_instead_of_comparison_operator)
+              .fixItReplace(Tok.getLoc(), "=");
```
Is it possible for the compiler to check this earlier and recover by parsing the expression normally?

vguerra replied:
The check could indeed be taken before l.272 for instance, but just to be sure I understand: in this case "recover" means that the token for '==' should be consumed so that the code can continue parsing the default value and so on?

jrose-apple replied:
Yeah, I was thinking you'd essentially have the compiler just pretend that the user wrote `=`.

vguerra replied:
Got it. Will try this.

rintaro replied:
I think this check and recovery should be performed *later*, like l.385.

vguerra replied:
but somehow I need to avoid the parsing logic to get into any of this [branches](https://github.com/apple/swift/blob/04ca9752cf4b5e2700b6663d0ddd381ea0fed3ab/lib/Parse/ParsePattern.cpp#L273-L376), so a check would be needed before.

rintaro replied:
It should work as expected if you consume `==` as if it's `=` and parse the default value after `==`.

vguerra replied:
pushed https://github.com/apple/swift/pull/26139/commits/d82dd84592c05655faa3e1e8fde3d6b038a59225 so check is now performed later as @rintaro suggested and compiler is recovering :).

vguerra committed:
Moving check for '==' close to that of '=' and diagnose there.

As well, parser recovers from usage of '==' and parses the whole
parameter clause.

vguerra commented on lib/Parse/ParsePattern.cpp:
```diff
@@ -67,10 +67,10 @@ void Parser::DefaultArgumentInfo::setFunctionContext(
   SyntaxParsingContext DefaultArgContext(P.SyntaxContext,
                                          SyntaxKind::InitializerClause);
-  SourceLoc equalLoc = P.consumeToken(tok::equal);
+  SourceLoc equalLoc = P.consumeToken(assignmentTok);
```
I wonder if consumption of either '=' or '==' should be taken out of here and be performed on caller site?

jrose-apple replied:
`consumeToken` also allows not passing a token kind at all, but I still think you might be right. It's weird to consume something when you don't know what it is.

rintaro replied:
I'm OK with:
```cpp
assert(Tok.is(tok::equal) ||
       (Tok.isBinaryOperator() && Tok.getText() == "=="));
consumeToken();
```
like
https://github.com/apple/swift/blob/26c4cccb4b971abae950d6a183354f70b2e3c0c7/lib/Parse/ParseDecl.cpp#L6632-L6634

vguerra replied:
thanks for the feedback .. address in https://github.com/apple/swift/pull/26139/commits/2cfe8ad52a0f2162808381101b6b80cb75a2b118

vguerra committed:
Addressing feedback on how `parseDefaultArgument` should consume '=' or '=='

rintaro commented:
@swift-ci Please smoke test

rintaro commented:
@swift-ci Please smoke test Linux platform

rintaro commented:
Thank you Victor!

vguerra commented:
thank you both for the guidance.
